dashboard/app/models/contact_rollups_raw.rb

```diff
@@ -22,22 +22,38 @@ def self.truncate_table
     ActiveRecord::Base.connection.execute("TRUNCATE TABLE #{table_name}")
   end
 
-  def self.extract_email_preferences
-    EmailPreference.all.find_in_batches(batch_size: 1000) do |email_preference_batch|
-      raw_contacts = []
-      email_preference_batch.each do |email_preference|
-        raw_contact = {
-          email: email_preference.email,
-          sources: "dashboard.#{EmailPreference.table_name}",
-          data: {opt_in: email_preference.opt_in},
-          data_updated_at: email_preference.updated_at
-        }
-        raw_contacts << raw_contact
-      end
+  # Generates insert query to run, and runs that query once constructed.
+  def self.extract_from_table(source_table, data_columns, email_column)
+    email_preferences_query = extract_from_table_query(source_table, data_columns, email_column)
+    ActiveRecord::Base.connection.execute(email_preferences_query)
+  end
+
+  # @param source_table [String] Table from which we want to extract data
+  # @param data_columns [Array] Columns we want reshaped into a single JSON object
+  # @param email_column [String] Column in source table we want to insert ino the email column
+  # @return [String] A SQL statement to extract and reshape data from the source table.
+  def self.extract_from_table_query(source_table, data_columns, email_column)
+    # TO DO: what to do when an email appears multiple times in the same source?
+    <<~SQL
+      INSERT INTO #{ContactRollupsRaw.table_name} (email, sources, data, data_updated_at, created_at, updated_at)
+      SELECT
+        #{email_column},
+        'dashboard.#{source_table}' AS sources,
+        #{extract_columns_into_mysql_json(data_columns)} AS data,
+        #{source_table}.updated_at AS data_updated_at,
+        NOW() AS created_at,
+        NOW() AS updated_at
+      FROM #{source_table}
+      WHERE #{email_column} IS NOT NULL AND #{email_column} != ''
+    SQL
+  end
+
+  # @param columns [Array] Column names to reshape
+  # @return [String] MySQL JSON_OBJECT() syntax for insertion
+  def self.extract_columns_into_mysql_json(columns)
+    return 'NULL' if columns.empty?
 
-      # ben to do: handle if required fields are missing, or if there's a duplicate?
-      # currently will fail if any required fields are missing
-      import! raw_contacts
-    end
+    last = columns.pop
+    'JSON_OBJECT(' + columns.map {|column| "'#{column}', #{column}, "}.join + "'#{last}', #{last}" + ')'
```

Comment:
You can do this without using `last`
```suggestion
'JSON_OBJECT(' + columns.map {|col| "'#{col}',#{col}"}.join(',') + ')'
```